src/main/GCNode.cpp

```diff
@@ -332,7 +332,18 @@ void GCNode::Marker::operator()(const GCNode* node)
     // Update mark  Beware ~ promotes to unsigned int.
     node->m_rcmms &= static_cast<unsigned char>(~s_mark_mask);
     node->m_rcmms |= s_mark;
-    ++m_marks_applied;
+    node->visitReferents(this);
+}
+
+void GCNode::CountingMarker::operator()(const GCNode* node)
+{
+    if (node->isMarked()) {
+	return;
+    }
+    // Update mark  Beware ~ promotes to unsigned int.
+    node->m_rcmms &= static_cast<unsigned char>(~s_mark_mask);
```

Comment:
Sure, but duplicating the 'if' is better than duplicating the whole thing.  The compiler will probably optimize out the 'if' anyway if you're concerned about that.

The syntax goes like:
    Marker::operator()(node);

It's not a virtual call since we've explicitly said which version we want to call, so it'll end up completely inlined.  It doesn't really matter though, since this code is semi-randomly jumping all over the heap, so I'd expect the performance to be entirely dominated by cache misses.